### PR TITLE
new channel for authentication

### PR DESCRIPTION
**Pull request by Nazianzenov**
* authentication channel created at the construction of any new LAO

### CODE REVIEW

File: be1-go/channel/authentication/mod.go [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+
+	data, ok := msgData.(*messagedata.AuthenticateUser)
+	if !ok {
+		return xerrors.Errorf("message %v isn't a authentication message", msgData)
+	}
+
+	// verify signatures and validity of the pop Token.
+	err := c.verifyAuthMessage(msg, *data)
+	if err != nil {
+		c.log.Err(err).Msg("Error: invalid authentication message received")
```
Comment (MariemBaccari) [inline be1-go/channel/authentication/mod.go:182]: Is it necessary to both log and return an error here ?

**Review by MariemBaccari — APPROVED** [review]
The code is clean. I only have a few small comments but it looks good to me.

File: be1-go/channel/authentication/mod.go [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+	failedToDecodeData = "failed to decode message data: %v"
+
+	//expiration time for the ID Token
+	expireDuration = time.Hour
+
+	// type of token in the popcha protocol
+	tokenBearer = "bearer"
+)
+
+// opaque string referencing a pop token for a given client ID
```
[the remark this answers is not in the document]
Comment (Nazianzenov) [reply be1-go/channel/authentication/mod.go:39]: it means that it is a pseudorandom values, its value itself doesn't have a meaning and is not attached to anything

File: be1-go/channel/authentication/mod.go [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+
+// Channel is used to handle authentication messages.
+type Channel struct {
+	sockets channel.Sockets
+	inbox   *inbox.Inbox
+
+	// channel path lao_id/authentication
+	channelID string
+
+	//clientID to pop token to identifier map
```
[the remark this answers is not in the document]
Comment (Nazianzenov) [reply be1-go/channel/authentication/mod.go:53]: good catch ahah

File: be1-go/channel/authentication/mod.go (hunk 1 above)
Comment (Nazianzenov) [reply be1-go/channel/authentication/mod.go:182]: not really, it was more for debug purposes, i'll change that!

File: be1-go/channel/authentication/mod.go [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+		Str(msgID, strconv.Itoa(catchup.ID)).
+		Msg("received a catchup")
+
+	return c.inbox.GetSortedMessages()
+}
+
+// Broadcast is forbidden, as authentication messages must be kept secret.
+func (c *Channel) Broadcast(_ method.Broadcast, _ socket.Socket) error {
+	panic("Broadcasting is not allowed in the authentication channel")
+}
```
[the remark this answers is not in the document]
Comment (Nazianzenov) [reply be1-go/channel/authentication/mod.go:146]: It is my bad, I didn't take care of that and forgot about it. Catch ups should be forbidden. However, i'm not sure about the use of panic in any of those places. I don't want the server to crash, but simply to not allow the use of those methods.

**Review by MariemBaccari — APPROVED** [review]
LGTM

**Review by Anghul — APPROVED** [review]
LGTM, just a short comment

**Pull request merged**